pkg/openstack/loadbalancer.go

```diff
@@ -306,6 +310,73 @@ func (s *floatingSubnetSpec) Matcher(tag bool) (matcher, error) {
 	return match, nil
 }
 
+func initGetMaxOctaviaAPIVersion() func(client *gophercloud.ServiceClient) (*semver.Version, error) {
+	// Why all the timing cache hassle? The underlying OpenStack cloud might get updated at some point
+	// and we probably want to notice that.
+	var cachedVersion *semver.Version = nil
+	var cachedTime time.Time
+
+	return func(client *gophercloud.ServiceClient) (*semver.Version, error) {
+		if cachedTime.Before(cachedTime.Add(time.Hour)) && cachedVersion != nil {
+			return cachedVersion, nil
+		}
+
+		allPages, err := apiversions.List(client).AllPages()
+		if err != nil {
+			return nil, err
+		}
+
+		apiVersions, err := apiversions.ExtractAPIVersions(allPages)
+		if err != nil {
+			return nil, err
+		}
+
+		var max *semver.Version = nil
+		for _, apiVersion := range apiVersions {
+			ver, err := semver.NewVersion(apiVersion.ID)
+
+			if err != nil {
+				// We're ignoring the error, if Octavia is returning anything odd we don't care.
+				klog.Warningf("Error when parsing Octavia API version %s: %v. Ignoring it", apiVersion.ID, err)
+				continue
+			}
+
+			if max == nil || ver.GreaterThan(max) {
+				max = ver
+			}
+		}
+
+		if max == nil {
+			// If we have max == nil at this point, then we couldn't read the versions at all.
+			// This happens for 2.0 API and let's return that.
+			max = semver.MustParse("v2.0")
+		}
+
+		klog.V(2).Infof("Detected Octavia API v%s", max)
+
+		cachedVersion = max
+		cachedTime = time.Now()
+		return max, nil
+	}
+}
+
+var getMaxOctaviaAPIVersion = initGetMaxOctaviaAPIVersion()
+
+func (lbaas *LbaasV2) isOctaviaVersionSupported(client *gophercloud.ServiceClient, constraint string) (bool, error) {
```

Comment:
Perhaps it's just me but I find this function hard to read.
How about something like that (not tested)?

```go
func (lbaas *LbaasV2) isOctaviaVersionSupported(client *gophercloud.ServiceClient, minimumVersion string) (bool, error) {
	remoteOctaviaVersion := semver.MustParse("2.0")
	if lbaas.opts.UseOctavia {
		var err error
		remoteOctaviaVersion, err = getMaxOctaviaAPIVersion(client)
		if err != nil {
			return false, err
		}
	}

	constraint, err := semver.NewConstraint(">= " + minimumVersion)
	if err != nil {
		return false, err
	}

	valid, _ := constraint.Validate(remoteOctaviaVersion)

	return valid, nil
}
```

Reply:
I actually have a PR to remove non-Octaiva case .. it will make those easier
but never got review and merge..

Reply:
I'll just use the version from util module instead.